colexec: optimize the read path a bit and cleanup internal memory handling

yuzefovich opened the pull request:
**colexec: reuse the same eval context for objects of the same proc spec**

In the row-execution processors we create a copy of the eval context for
each processor core because the processor might modify it. In the
vectorized engine we took it to the extreme - we're currently creating
a copy for every usage of the eval context. That is unnecessary, and
this commit makes the vectorized operator creator to be more like the
row engine - by reusing the same eval context for all operators that are
created for a single processor spec.

This commit also switches to using the "original" eval context that is
part of the flow context when instantiating a ColumnFactory and when
initializing materializers because those two components don't modify the
eval context.

Release note: None

**colconv,colfetcher: pool allocations of converters and cTableInfos**

This commit pools the allocations of the converters when they are used
in the materializers as well as `cTableInfo` structs used by the
ColBatchScan. Additionally, this commit reuses the same ColumnFactory
when setting up the whole flow and removes a copy of a processor spec
when checking whether it is supported.

Another notable change is reusing the same global empty post-processing
spec when creating the materializers. It is thread-safe because the object
is never modified.

Release note: None

**colbuilder: remove redundant expr deserialization after scan**

Whenever we're creating a `ColBatchScan`, we're initializing
a `ProcOutputHelper` in order to obtain the set of needed columns (based
on the schema of the table and the post-processing spec). As part of
that initialization we're obtaining well-typed expression for everything
in the post-processing spec. However, later when actually planning the
post-processing in the vectorized engine we deserialize the expressions
again. This is redundant, and the code has been adjusted to reuse the
same helper for both scan operator creation and post-processing
planning. Additionally, those `ProcOutputHelper` objects are now pooled.

Release note: None

**colexec: update the allocator with the internal memory usage**

This commit removes `InternalMemoryOperator` interface that was put
in-place for the operators to register their static memory usage. This
was introduced before we had the allocator object (I think), and now we
can remove that interface and simply update the allocators with the
corresponding internal memory usage. There were only two operators that
actually implemented the interface, and both of them were converted to
the new pattern.

This commit also removes the separation of streaming memory accounts in
the vectorized flow setup and merges them with all other accounts.

Release note: None

yuzefovich committed:
colexec: reuse the same eval context for objects of the same proc spec

In the row-execution processors we create a copy of the eval context for
each processor core because the processor might modify it. In the
vectorized engine we took it to the extreme - we're currently creating
a copy for every usage of the eval context. That is unnecessary, and
this commit makes the vectorized operator creator to be more like the
row engine - by reusing the same eval context for all operators that are
created for a single processor spec.

This commit also switches to using the "original" eval context that is
part of the flow context when instantiating a ColumnFactory and when
initializing materializers because those two components don't modify the
eval context.

Release note: None

yuzefovich committed:
colexec: update the allocator with the internal memory usage

This commit removes `InternalMemoryOperator` interface that was put
in-place for the operators to register their static memory usage. This
was introduced before we had the allocator object (I think), and now we
can remove that interface and simply update the allocators with the
corresponding internal memory usage. There were only two operators that
actually implemented the interface, and both of them were converted to
the new pattern.

This commit also removes the separation of streaming memory accounts in
the vectorized flow setup and merges them with all other accounts.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/56224)
<!-- Reviewable:end -->

yuzefovich commented:
Firth four commits are in separate PRs, the benchmarks of the other commits:
```
name                                           old time/op    new time/op    delta
FlowSetup/vectorize=true/distribute=true-24      19.0µs ± 5%    18.3µs ± 5%  -3.93%  (p=0.005 n=10+10)
FlowSetup/vectorize=true/distribute=false-24     16.7µs ± 3%    15.9µs ± 1%  -4.80%  (p=0.000 n=10+9)
FlowSetup/vectorize=false/distribute=true-24     16.1µs ± 3%    16.1µs ± 3%    ~     (p=0.842 n=9+10)
FlowSetup/vectorize=false/distribute=false-24    15.1µs ± 3%    15.2µs ± 1%    ~     (p=0.315 n=10+9)

name                                           old alloc/op   new alloc/op   delta
FlowSetup/vectorize=true/distribute=true-24      39.2kB ± 0%    36.0kB ± 0%  -8.11%  (p=0.000 n=10+9)
FlowSetup/vectorize=true/distribute=false-24     35.7kB ± 0%    32.5kB ± 0%  -8.87%  (p=0.000 n=10+9)
FlowSetup/vectorize=false/distribute=true-24     33.4kB ± 0%    33.4kB ± 0%  -0.10%  (p=0.000 n=9+8)
FlowSetup/vectorize=false/distribute=false-24    30.0kB ± 0%    30.0kB ± 0%  -0.15%  (p=0.027 n=10+9)

name                                           old allocs/op  new allocs/op  delta
FlowSetup/vectorize=true/distribute=true-24         350 ± 0%       340 ± 0%  -2.86%  (p=0.000 n=10+9)
FlowSetup/vectorize=true/distribute=false-24        319 ± 0%       309 ± 0%  -3.13%  (p=0.000 n=9+10)
FlowSetup/vectorize=false/distribute=true-24        286 ± 0%       284 ± 0%  -0.70%  (p=0.000 n=10+10)
FlowSetup/vectorize=false/distribute=false-24       256 ± 1%       254 ± 1%  -0.82%  (p=0.008 n=10+10)
```

yuzefovich committed:
colbuilder: remove redundant expr deserialization after scan

Whenever we're creating a `ColBatchScan`, we're initializing
a `ProcOutputHelper` in order to obtain the set of needed columns (based
on the schema of the table and the post-processing spec). As part of
that initialization we're obtaining well-typed expression for everything
in the post-processing spec. However, later when actually planning the
post-processing in the vectorized engine we deserialize the expressions
again. This is redundant, and the code has been adjusted to reuse the
same helper for both scan operator creation and post-processing
planning. Additionally, those `ProcOutputHelper` objects are now pooled.

Release note: None

yuzefovich commented:
Okay, I cleaned up this PR and merged its dependencies, so it is RFAL.

asubiotto reviewed (changes requested):
> Reviewed 17 of 17 files at r1, 9 of 9 files at r2, 3 of 3 files at r3, 17 of 17 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56224)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/colbuilder/execplan.go, line 612 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56224#-MLRvmZCC0Z0oTWla1Gb:-MLRvmZCC0Z0oTWla1Gc:bx7e5n2) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b559976b58e6350c205e6d4d25fe1390d71ec95/pkg/sql/colexec/colbuilder/execplan.go#L612)):*
> ```Go
> 	ctx context.Context,
> 	flowCtx *execinfra.FlowCtx,
> 	evalCtx *tree.EvalContext,
> ```

Why pass in `evalCtx` instead of doing `evalCtx := flowCtx.NewEvalCtx()` immediately? It avoids having to put it in every `NewColOperator` callsite.
___
*[pkg/sql/colexec/colbuilder/execplan.go, line 1317 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56224#-MLS-0JT1jpItDBOZLxa:-MLS-0JU81rrtUjrI8SP:bez6c8s) ([raw file](https://github.com/cockroachdb/cockroach/blob/44200eca8ef429e38f3236f41598f4761b328866/pkg/sql/colexec/colbuilder/execplan.go#L1317)):*
> ```Go
> 	post *execinfrapb.PostProcessSpec,
> 	factory coldata.ColumnFactory,
> 	scanHelper *execinfra.ProcOutputHelper,
> ```

Rename and reword comment to make this a a more general proc output helper to be used if provided, rather than specific to scans.
___
*[pkg/sql/colflow/vectorized_flow.go, line 473 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56224#-MLRzKJI6UQ3b5znaPzW:-MLRzKJI6UQ3b5znaPzX:b30qf9v) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b559976b58e6350c205e6d4d25fe1390d71ec95/pkg/sql/colflow/vectorized_flow.go#L473)):*
> ```Go
> 	// streamingMemAccounts contains all memory accounts of the non-buffering
> 	// components in the vectorized flow.
> 	streamingMemAccounts []*mon.BoundAccount
> ```

I don't think I see a mention of the streamingMemAccounts removal in the commit message, could you add that?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56224)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/colbuilder/execplan.go, line 612 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56224#-MLRvmZCC0Z0oTWla1Gb:-MLU5d5CAmj804TycpgJ:bnhdeyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b559976b58e6350c205e6d4d25fe1390d71ec95/pkg/sql/colexec/colbuilder/execplan.go#L612)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Why pass in `evalCtx` instead of doing `evalCtx := flowCtx.NewEvalCtx()` immediately? It avoids having to put it in every `NewColOperator` callsite.
</blockquote></details>

Good point, fixed. Not sure why I didn't think of it myself.
___
*[pkg/sql/colexec/colbuilder/execplan.go, line 1317 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/56224#-MLS-0JT1jpItDBOZLxa:-MLU78eY2c1X4gd02UMr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44200eca8ef429e38f3236f41598f4761b328866/pkg/sql/colexec/colbuilder/execplan.go#L1317)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Rename and reword comment to make this a a more general proc output helper to be used if provided, rather than specific to scans.
</blockquote></details>

Done.
___
*[pkg/sql/colflow/vectorized_flow.go, line 473 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/56224#-MLRzKJI6UQ3b5znaPzW:-MLU67A3Df6RCrs8eTQe:b-vivvgz) ([raw file](https://github.com/cockroachdb/cockroach/blob/9b559976b58e6350c205e6d4d25fe1390d71ec95/pkg/sql/colflow/vectorized_flow.go#L473)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I don't think I see a mention of the streamingMemAccounts removal in the commit message, could you add that?
</blockquote></details>

Yeah, I added it last minute and forgot to update the commit message, fixed.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
colconv,colfetcher: pool allocations of converters and cTableInfos

This commit pools the allocations of the converters when they are used
in the materializers as well as `cTableInfo` structs used by the
ColBatchScan. Additionally, this commit reuses the same ColumnFactory
when setting up the whole flow and removes a copy of a processor spec
when checking whether it is supported.

Another notable change is reusing the same global empty post-processing
spec when creating the materializers. It is thread-safe because the
object is never modified.

Release note: None

yuzefovich commented:
I updated the microbenchmarks with recent changes and reran KV95 benchmark on GCE and AWS with different durations (full output is [here](https://gist.github.com/yuzefovich/981ab48a373bedb9db82b621fb388c6a), "old" is `vectorize_row_count_threshold=1000` and "new" is `0`).
- GCE
  - 1 min ramp, 5 min duration: **-3.48%**
  - 3 min ramp, 15 min duration: **-2.46%**
  - 5 min ramp, 30 min duration: **-3.44%**
  - 10 min ramp, 60 min duration: **+0.28%**
- AWS
  - 1 min ramp, 5 min duration: **-0.64%**
  - 3 min ramp, 15 min duration: **-2.25%**
  - 5 min ramp, 30 min duration: **-2.14%**
  - 10 min ramp, 60 min duration: **-1.70%**

At this point I don't know what else we can optimize. The only remaining idea I want to try is Alfonso's suggestion on removing the duplication of calls to `supportedNatively` and `canWrap` when we have a single local flow by introducing an ability to fallback to row-based flow if the vectorized flow setup fails. However, I doubt it'll have much impact since `IsSupported` check doesn't show up in the profiles. As another next step I want to run TPCC benchmarks to see whether there is a change there, and in case there is a positive one I feel like we could make an argument for setting the threshold to 0 even with about 2 percent hit on KV95.

asubiotto reviewed (approved):
According to your investigation, what is the reason for the remaining performance difference? Speaking with no context: one thing we can try which might be a bit complex is pooling mem batches. The thing is we should probably make sure that allocations are the problem here. One way to do this is also to pull up the execution trace, which has a view of the heap and how often GC kicks in, but I can't seem to find it in the admin UI any more.

This PR <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 36 of 36 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56224#-:-MLgZOljB0_VtfG0i7sZ:b-p762p8)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
As discussed in our one-on-one, the vectorized flow setup is a bit more expensive, and probably the largest difference is that we need to create two components (ColBatchScan and materializer) instead of one (tableReader). I'll look into several ideas we've talked about and will open up a separate PR for those, so I'll go ahead and merge this as is.

TFTR!

bors r+
